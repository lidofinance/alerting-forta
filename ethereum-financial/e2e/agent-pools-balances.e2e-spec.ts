import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-pools-balances e2e tests", () => {
  let runBlock: (
    blockHashOrNumber: string | number,
    initBlock?: number
  ) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(
        provideAgentPath("subagents/pools-balances/agent-pools-balances")
      ),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process block with imbalanced Curve pool",
    async () => {
      const findings = await runBlock(16804419);
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with Curve pool rapid imbalance change",
    async () => {
      const findings = await runBlock(16306452, 16300452);
      expect(
        findings
          .filter(
            (finding) => finding.alertId == "CURVE-POOL-IMBALANCE-RAPID-CHANGE"
          )
          .at(0)
      ).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with significant Curve pool change",
    async () => {
      const findings = await runBlock(16870590, 16870589);
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with imbalanced Balancer pool",
    async () => {
      const findings = await runBlock(16398269, 16347860);
      expect(
        findings
          .filter((finding) => finding.alertId == "BALANCER-POOL-IMBALANCE")
          .at(0)
      ).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with significant Balancer pool change",
    async () => {
      const findings = await runBlock(17000731, 17000000);
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with Balancer pool rapid imbalance change",
    async () => {
      const findings = await runBlock(16306450, 16306009);
      expect(
        findings
          .filter(
            (finding) =>
              finding.alertId == "BALANCER-POOL-IMBALANCE-RAPID-CHANGE"
          )
          .at(0)
      ).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with low stETH PEG on Chainlink",
    async () => {
      const findings = await runBlock(15485829, 15485828);
      expect(
        findings
          .filter((finding) => finding.alertId == "LOW-STETH-CHAINLINK-PEG")
          .at(0)
      ).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with decreased stETH PEG on Curve",
    async () => {
      const findings = await runBlock(16038267, 16037266);
      expect(
        findings
          .filter((finding) => finding.alertId == "STETH-CURVE-PEG-DECREASE")
          .at(0)
      ).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );
});
