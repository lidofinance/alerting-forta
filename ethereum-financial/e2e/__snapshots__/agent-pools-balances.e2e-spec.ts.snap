// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-pools-balances e2e tests should process block with Balancer pool rapid imbalance change 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "BALANCER-POOL-IMBALANCE-RAPID-CHANGE",
  "description": "Prev reported pool sate:
ETH - 62886.94 (39.89%)
stETH - 94771.45 (60.11%)
Current pool state:
ETH - 54704.67 (34.84%)
wstETH - 102297.29 (65.16%)",
  "labels": Array [],
  "metadata": Object {},
  "name": "üö® Balancer Pool rapid imbalance change",
  "protocol": "ethereum",
  "severity": 4,
  "type": 2,
}
`;

exports[`agent-pools-balances e2e tests should process block with Curve pool rapid imbalance change 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "CURVE-POOL-IMBALANCE-RAPID-CHANGE",
  "description": "Prev reported pool sate:
ETH - 238192.73 (37.39%)
stETH - 398801.52 (62.61%)
Current pool state:
ETH - 207068.42 (32.25%)
stETH - 435042.52 (67.75%)",
  "labels": Array [],
  "metadata": Object {},
  "name": "üö® Curve Pool rapid imbalance change",
  "protocol": "ethereum",
  "severity": 4,
  "type": 2,
}
`;

exports[`agent-pools-balances e2e tests should process block with decreased stETH PEG on Curve 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "STETH-CURVE-PEG-DECREASE",
  "description": "stETH PEG on Curve decreased to 0.9704",
  "labels": Array [],
  "metadata": Object {
    "peg": "0.9704",
  },
  "name": "‚ö†Ô∏è stETH PEG on Curve decreased",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-pools-balances e2e tests should process block with imbalanced Balancer pool 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "BALANCER-POOL-IMBALANCE",
  "description": "Current pool state:
ETH - 41075.27 (44.16%)
wstETH - 51934.46 (55.84%)",
  "labels": Array [],
  "metadata": Object {},
  "name": "‚ö†Ô∏è Balancer Pool is imbalanced",
  "protocol": "ethereum",
  "severity": 4,
  "type": 2,
}
`;

exports[`agent-pools-balances e2e tests should process block with imbalanced Curve pool 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "CURVE-POOL-IMBALANCE",
  "description": "Current pool state:
ETH - 432969.66 (44.87%)
stETH - 531933.72 (55.13%)",
  "labels": Array [],
  "metadata": Object {},
  "name": "‚ö†Ô∏è Curve Pool is imbalanced",
  "protocol": "ethereum",
  "severity": 4,
  "type": 2,
}
`;

exports[`agent-pools-balances e2e tests should process block with significant Balancer pool change 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "BALANCER-POOL-SIZE-CHANGE",
  "description": "Balancer Pool size has increased by 5.26% since the last block",
  "labels": Array [],
  "metadata": Object {
    "sizeAfter": "159703765204466775027621",
    "sizeBefore": "151721475467519582335566",
  },
  "name": "‚ö†Ô∏è Significant Balancer Pool size change",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-pools-balances e2e tests should process block with significant Curve pool change 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "CURVE-POOL-SIZE-CHANGE",
  "description": "Curve Pool size has increased by 15.26% since the last block",
  "labels": Array [],
  "metadata": Object {
    "sizeAfter": "960964420322769261730507",
    "sizeBefore": "833732119622588732317889",
  },
  "name": "‚ö†Ô∏è Significant Curve Pool size change",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;
